index.bs

```diff
@@ -3488,6 +3478,45 @@ WritableStreamDefaultController(<var>stream</var>, <var>underlyingSink</var>, <v
   1. Perform ! WritableStreamDefaultControllerError(*this*, _e_).
 </emu-alg>
 
+<h4 id="ws-default-controller-internal-methods">Writable Stream Default Controller Internal Methods</h4>
+
+The following are additional internal methods implemented by each {{WritableStreamDefaultController}} instance. The
+writable stream implementation will call into these.
+
+<p class="note">The reason these are in method form, instead of as abstract operations, is to make it clear that the
+writable stream implementation is decoupled from the controller implementation, and could in the future be expanded with
+other controllers, as long as those controllers implemented such internal methods. A similar scenario is seen for
+readable streams, where there actually are multiple controller types and as such the counterpart internal methods are
+used polymorphically.
+
+<h5 id="ws-default-controller-private-abort" oldids="writable-stream-default-controller-abort">\[[AbortSteps]]()</h5>
+
+<emu-alg>
+  1. Perform ! ResetQueue(*this*).
+  1. Let _sinkAbortPromise_ be ! PromiseInvokeOrNoop(*this*.[[underlyingSink]], `"abort"`, « _reason_ »).
+  1. Return the result of <a>transforming</a> _sinkAbortPromise_ with a fulfillment handler that returns
+     *undefined*.
+</emu-alg>
+
+<h5 id="ws-default-controller-private-start" oldids="writable-stream-default-controller-start">\[[StartSteps]]()</h5>
+
+<emu-alg>
+  1. Let _startResult_ be ? InvokeOrNoop(*this*.[[underlyingSink]], `"start"`, « *this* »).
+  1. Let _stream_ be *this*.[[controlledWritableStream]].
+  1. Let _startPromise_ be <a>a promise resolved with</a> _startResult_:
+    1. <a>Upon fulfillment</a>  of _startPromise_,
+      1. Set *this*.[[started]] to *true*.
+      1. If _stream_.[[state]] is `"errored"`, perform ! WritableStreamRejectAbortRequestIfPending(_stream_).
+      1. Otherwise, perform ! WritableStreamHandleAbortRequestIfPending(_stream_).
+      1. Perform ! WritableStreamDefaultControllerAdvanceQueueIfNeeded(*this*).
+    1. <a>Upon rejection</a> of _startPromise_ with reason _r_,
+      1. Assert: _stream_.[[state]] is `"writable"` or `"errored"`.
+      1. Perform ! WritableStreamDefaultControllerErrorIfNeeded(*this*, _r_).
+      1. Perform ! WritableStreamRejectAbortRequestIfPending(_stream_).
+</emu-alg>
+
+
+
```

Comment:
Three blank lines is unusual in index.bs (I think this is the only place with 3).